tfx/experimental/pipeline_testing/pipeline_recorder_utils.py

```diff
@@ -0,0 +1,157 @@
+  """Returns a dictionary holding list of executions for all run_id in MLMD.
+
+  Args:
+    metadata_connection: Instance of metadata.Metadata for I/O to MLMD.
+
+  Returns:
+    A dictionary that holds list of executions for a run_id.
+  """
+  execution_dict = defaultdict(list)
+  for execution in metadata_connection.store.get_executions():
+    execution_run_id = execution.properties['run_id'].string_value
+    execution_dict[execution_run_id].append(execution)
+  return execution_dict
+
+def _get_latest_executions(metadata_connection: metadata.Metadata,
+                           pipeline_name: Text
+                           ) -> List[metadata_store_pb2.Execution]:
+  """Fetches executions associated with the latest context.
+
+  Args:
+    metadata_connection: Instance of metadata.Metadata for I/O to MLMD.
+    pipeline_name: Name of the pipeline to rerieve the latest executions for.
+
+  Returns:
+    List of executions for the latest run of a pipeline with the given
+    pipeline_name.
+  """
+  pipeline_run_contexts = [
+      c for c in metadata_connection.store.get_contexts_by_type(
+          metadata._CONTEXT_TYPE_PIPELINE_RUN)  # pylint: disable=protected-access
+      if c.properties['pipeline_name'].string_value == pipeline_name
+  ]
+  latest_context = max(pipeline_run_contexts,
+                       key=lambda c: c.last_update_time_since_epoch)
+  return metadata_connection.store.get_executions_by_context(latest_context.id)
+
+def record_pipeline(output_dir: Text,
+                    metadata_db_uri: Optional[Text],
+                    host: Optional[Text],
+                    port: Optional[int],
+                    pipeline_name: Optional[Text],
+                    run_id: Optional[Text]) -> None:
+  """Record pipeline run with run_id to output_dir. For the beam pipeline,
+  metadata_db_uri is required. For KFP, host and port should be specified.
+  If run_id is not specified, then pipeline_name ought to be specified in
+  order to fetch the latest execution for the specified pipeline.
+
+  Args:
+    output_dir: Directory path where the pipeline outputs should be recorded.
+    metadata_db_uri: Uri to metadata db.
+    host: Hostname of the metadata grpc server
+    port: Port number of the metadata grpc server.
+    pipeline_name: Pipeline name, which is required if run_id isn't specified.
+    run_id: Pipeline execution run_id.
+
+  Raises:
+    ValueError: In cases of invalid arguments:
+      - metadata_db_uri is None or host and/or port is None.
+      - run_id is None and pipeline_name is None.
+  """
+  if host is not None and port is not None:
+    metadata_config = metadata_store_pb2.MetadataStoreClientConfig(
+        host=host, port=port)
+  elif metadata_db_uri is not None:
+    metadata_config = metadata.sqlite_metadata_connection_config(
+        metadata_db_uri)
+  else:
+    raise ValueError("For KFP, host and port are required. "\
```

Comment:
Prefer not to use backslash for line continuation.

https://google.github.io/styleguide/pyguide.html#32-line-length